catboost/python-package/catboost/core.py

```diff
@@ -3451,6 +3451,137 @@ def randomized_search(self, param_distributions, X, y=None, cv=3, n_iter=10, par
+                              random_state=random_state)
+        best_params = {optimized_params_names[i]: results.x[i] for i in range(len(optimized_params_names))}
+        self.set_params(**best_params)
+        if refit:
+            self.set_params(**best_params)
+            self.fit(X, y, silent=True)
+        return best_params
+
+    def _convert_to_asymmetric_representation(self):
+        self._object._convert_oblivious_to_asymmetric()
+
+
+class SkoptObjective(object):
+    def __init__(self, X, y, random_state, train_size, cv,
+                 partition_random_seed, optimized_params_names,
+                 model, loss_function, train_pool, search_by_train_test_split,
+                 to_minimize_objective, const_params, init_params):
+        self.X = X
+        self.y = y
+        self.random_state = random_state
+        self.train_size = train_size
+        self.cv = cv
+        self.partition_random_seed = partition_random_seed
+        self.optimized_params_names = optimized_params_names
+        self.model = model
+        self.loss_function = loss_function
+        self.train_pool = train_pool
+        self.to_minimize_objective = to_minimize_objective
+        self.search_by_train_test_split = search_by_train_test_split
+        self.const_params = const_params
+        self.init_params = init_params
+
+    def __call__(self, params):
+        from sklearn.model_selection import train_test_split
+
+        params_dict = dict(zip(self.optimized_params_names, params))
+        if isinstance(self.model, CatBoostClassifier):
+            self.model = CatBoostClassifier(**self.init_params)
+        elif isinstance(self.model, CatBoostRegressor):
+            self.model = CatBoostRegressor(**self.init_params)
+        elif isinstance(self.model, CatBoost):
+            self.model = CatBoost(**self.init_params)
+
+        self.model.set_params(**self.const_params)
+        self.model.set_params(**params_dict)
+        if self.search_by_train_test_split:
+            if isinstance(self.X, Pool):
+                self.X = self.X.get_features()
+                self.y = self.X.get_label()
+            X_train, X_val, y_train, y_val = train_test_split(self.X,
+                                                              self.y,
+                                                              random_state=self.partition_random_seed,
+                                                              train_size=self.train_size)
+            self.model.fit(X_train, y_train, silent=True, eval_set=(X_val, y_val))
+            result = self.model.get_best_score()['validation'][self.loss_function]
+        else:
+            result = self.model.cv(self.train_params["train_pool"],
+                              params=params_dict,
+                              fold_count=self.cv,
+                              partition_random_seed=self.partition_random_seed)
+            result = list(result["test-" + self.loss_function + "-mean"])[-1]
```

Comment:
this will work not for all loss_functions, plus loss_function might be None in this code